src/Futhark/LSP/PositionMapping.hs

```diff
@@ -0,0 +1,78 @@
+-- | Provide mapping between position in stale content and current.
+module Futhark.LSP.PositionMapping
+  ( mappingFromDiff,
+    PositionMapping,
+    toStalePos,
+    toCurrentLoc,
+    StaleFile (..),
+  )
+where
+
+import Data.Algorithm.Diff (Diff, PolyDiff (Both, First, Second), getDiff)
+import Data.Bifunctor (Bifunctor (bimap, first, second))
+import qualified Data.Text as T
+import qualified Data.Vector as V
+import Futhark.Util.Loc (Loc (Loc), Pos (Pos))
+import Language.LSP.VFS (VirtualFile)
+
+-- | A mapping between current file content and the stale (last successful compiled) file content.
+-- The first component of the pair provides the mapping from stale position to current,
+-- e.g. old2new[2] = 4 means "line 2" in the stale file corresponds to "line 4" in the current file.
+-- The second component vice versa.
+-- Currently, only supports entire line mapping,
+-- more detailed mapping might be achieved via referring to haskell-language-server@efb4b94
+newtype PositionMapping = PositionMapping (V.Vector Int, V.Vector Int)
+  deriving (Show)
+
+-- | Stale text document stored in state.
+data StaleFile = StaleFile
+  { -- | The last successfully compiled file content.
+    -- Using VirtualFile for convenience, we can use anything with {version, content}
+    staleContent :: VirtualFile,
+    -- | PositionMapping between current and stale file content.
+    -- Nothing if last type-check is successful.
+    staleMapping :: Maybe PositionMapping
+  }
+
+instance Show StaleFile where
+  show (StaleFile _ mapping) = show mapping
```

Comment:
Non-`derived` instances of `Show` are a bad idea.  Let people write `show . staleMapping` if they want to see that part.